### PR TITLE
Allow editing of new menu items from the block inspector

### LINKED ISSUE #20748: Navigation menu block: allow editing of new menu items from the block inspector UI
A [suggestion](https://github.com/WordPress/gutenberg/issues/20742#issuecomment-596773200) from #20742.

The block navigator in the navigation menu block has proven to be a familiar and easy way to navigate UI, particularly for users who are using only a keyboard.

76255145-4d5c0b80-6213-11ea-876a-7d02a83d8bbd

Currently you can add new links from it, but one missing piece that would make it much more useful is to also be able to edit new menu items (name, URL, etc) right from the block navigator.

### PR DESCRIPTION
**Pull request by adamziel**
## Description

This is a follow-up to https://github.com/WordPress/gutenberg/pull/21948
Fixes https://github.com/WordPress/gutenberg/issues/20748.

It plugs RichText editor into those slots `navigation-link/edit.js`. As a result, it's possible to edit the label from the `BlockNavigationList` itself.

## How has this been tested?

1. Apply this PR and https://github.com/WordPress/gutenberg/pull/22355

### On experimental nav-menus page:

1. Go to `/wp-admin/admin.php?page=gutenberg-navigation`
1. Edit the text from the navigation structure panel like on the gif below:

1. Try with both top-level and nested navigation items

### In posts editor:

1. Go to posts > add new
1. Add navigation block with some nested items
1. Try editing titles through the navigator in block inspector (slow! performance fix: #21973):
Zrzut ekranu 2020-04-29 o 13 35 11

1. Make sure to use rich formatting like pressing enter or using keyboard shortcuts (e.g. cmd+b on mac to bold selected text).

1. Try editing titles through the navigator in modal:
Zrzut ekranu 2020-04-29 o 13 35 17

1. Confirm global block navigation remains non-editable and only allows block selection on click:

Zrzut ekranu 2020-04-29 o 13 36 15

## Types of changes
Bug fix (non-breaking change which fixes an issue)

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

File: packages/block-library/src/navigation-link/edit.js [hunk 1]
```diff
@@ -193,6 +222,14 @@ function NavigationLinkEdit( {
...
 				</PanelBody>
 			</InspectorControls>
+			<BlockNavigationListItemFill>
+				<BlockNavigationListItem
+					wrapperComponent="div"
+					onClick={ noop }
+				>
+					{ editField }
+				</BlockNavigationListItem>
+			</BlockNavigationListItemFill>
```
Comment (talldan) [inline packages/block-library/src/navigation-link/edit.js:232]: Given the way the design is going, I'm wondering if we should consider a more straightforward component that only makes it possible to edit text. Internally I think this would still use the fill, but externally it'd just be a single component that makes it possible edit the text for a block:
```jsx
// Not sure on the naming here, but to illustrate:
<BlockNavigationEdit
  onChange={ ( labelValue ) => setAttributes( { label: labelValue } ) }
  value={ value }
/>
```

Explaining my thought process—originally when we looked at using a slot/fill I wasn't sure if there would be additional functionality built in to a single block in the block navigation—for example, being able to edit the href of a navigation link. That doesn't seem to be the case now, any functionality like that (if it is added) seems like it'll end up in the dropdown menu shown in https://github.com/WordPress/gutenberg/issues/22089.

I think there's a benefit in that this gives more control over how the block navigation will look and what features are exposed (for example, more control over handling formatting).

What do you think?

**Comment by draganescu** [comment]
I tested this and it works all right, the performance in the editor is good even with larger menus.

File: packages/block-editor/src/components/block-styles/index.js [hunk 2]
```diff
@@ -77,7 +77,7 @@ const useGenericPreviewBlock = ( block, type ) =>
 						innerBlocks: type.example.innerBlocks,
 				  } )
 				: cloneBlock( block ),
-		[ block, type ]
+		[ type.example ? block.name : block, type ]
```
Comment (draganescu) [inline packages/block-editor/src/components/block-styles/index.js:80]: 👋  is this change related to this PR?

File: packages/block-library/src/navigation-link/edit.js (hunk 1 above)
Comment (adamziel) [reply packages/block-library/src/navigation-link/edit.js]: @talldan yup it makes sense, I just adjusted the PR

File: packages/block-editor/src/components/block-styles/index.js (hunk 2 above)
Comment (adamziel) [reply packages/block-editor/src/components/block-styles/index.js:80]: It fixes the performance issue - I'll extract it into a separate PR

Comment (adamziel) [reply packages/block-editor/src/components/block-styles/index.js:80]: https://github.com/WordPress/gutenberg/pull/22355

**Comment by adamziel** [comment]
@draganescu would you mind approving if it works well? :-)

**Review by draganescu — APPROVED** [review]
Forgot to approve! Here you go @adamziel :shipit:
